File: jangwon/5주차/5-1/단어변환.js
// DFS 풀이 
let result = Infinity;
let check = [];

const diff = (word1, word2) => {
    let diffNumber = 0;
    for (let i = 0; i< word1.length; i++) {
        if (word1[i] !== word2[i]) diffNumber += 1;
    }
    
    return diffNumber === 1 ? true : false;
};

const dfs = (begin, target, words, depth, check) => {
  if (begin == target && result > depth) {
    result = depth;
    return;
  } else {
      for (let i = 0; i < words.length; i++) {
        if (!check[i] && diff(begin, words[i])) {
          depth += 1;
          check[i] = true;
          dfs(words[i], target, words, depth, check);
          check[i] = false;
          depth -= 1;
        }
     }
  }
};

const solution = (begin, target, words) => {
    // 예외 처리
    if (!words.includes(target)) return 0;
    
    // 방문 함수 생성
    words.forEach((_, idx) => check[idx] = false);
    dfs(begin, target, words, 0, check);
    return result === Infinity ? 0 : result;
};

File: jangwon/5주차/5-1/단어변환.py
result = 100
check = []

# 단어 차이 판별 함수
def diffrent_Letter(word1, word2):
    diff = 0
    for idx, word in enumerate(word1):
        if word != word2[idx]: diff += 1
    
    return diff == 1

def dfs(begin, target, words, depth, check):
    if begin == target and result > depth:
        # 최소 값 바꿔 주기
        result = depth
        return
    else:
        for idx, start in enumerate(words):
            if diffrent_Letter(begin, start[idx]) and check[idx] == False:
                # 방문 처리
                check[idx] = True
                # 깊이 더 해주기
                depth += 1
                dfs(start, target, words, depth, check)
                # 백 트래킹이기 때문에 방문 해제 
                check[idx] = False
                # 다시 돌아갔기 때문에 depth 해제
                depth -= 1

def solution(begin, target, words):
    # 바꿀 단어가 하나도 없는 경우
    if begin not in words:
        return 0
    
    # 방문 배열 만들어 주기
    check = [False for i in range(len(words))]
    
    dfs(begin, target, words, depth, check)
    
    if result != 100:
        return result
    else:
        0